### PR TITLE
dataflow: tweak TAIL sink

### PR DESCRIPTION
**Pull request by benesch**
Make it clearer that the TAIL sink only operates on updates with a diff
of 1. Also, sort by timestamp without allocating a new vector.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5450)

### CODE REVIEW

**Review by ruchirK — APPROVED** [review]

**Review by cirego — COMMENTED** [review]
Why not allow for +N or -N for diff for `TAIL`? I wrote the wikirecent demo assuming that `diff` could be an integer multiple of the number of rows to insert or delete. It seems weird to limit TAIL like this?

Unless this is an explicit property of the interface that TAIL is consuming from?

File: src/dataflow/src/sink/tail.rs [hunk 1]
```diff
@@ -62,11 +58,11 @@ pub fn tail<G>(
...
-            // Sort results by time and convert to Vec<Row>. We use stable sort here even
-            // though it is slower because it will produce deterministic results since the
-            // cursor will always produce rows in the same order.
-            results.sort_by_key(|(time, _)| *time);
-            let mut results: Vec<Row> = results.into_iter().map(|(_, row)| row).collect();
+            // Sort results by time, which is the first column in the row. We
+            // use stable sort here even though it is slower because it will
+            // produce deterministic results since the cursor will always
+            // produce rows in the same order.
+            results.sort_by(|row1, row2| row1.unpack_first().cmp(&row2.unpack_first()));
```
Comment (umanwizard) [inline src/dataflow/src/sink/tail.rs:65]: I'm not sure about this, it again seems to be mixing properties of the envelope and the connector - we now rely on the fact that the envelope stuffs the timestamp into the first column of the data.

**Comment by ruchirK** [comment]
> Unless this is an explicit property of the interface that TAIL is consuming from?

The surrounding code in render/mod.rs turns arbitrary diffs into unary diffs with the old diff value placed into the row. 

e.g. it turns

```
(foo, t1, +5)
```

into

```
([foo, t1, 5], t1, +1)
```

File: src/dataflow/src/sink/tail.rs [hunk 2]
```diff
@@ -42,18 +42,14 @@ pub fn tail<G>(
                     while cursor.val_valid(&batch) {
                         let row = cursor.val(&batch);
                         cursor.map_times(&batch, |time, diff| {
-                            assert!(*diff >= 0, "negative multiplicities sinked in tail");
-                            let diff = *diff as usize;
+                            assert!(*diff == 1, "non-unary diff sinked in tail");
```
Comment (mjibson) [inline src/dataflow/src/sink/tail.rs:45]: I don't understand why this is the case. What is guaranteeing that the diff is always 1?

Comment (mjibson) [reply src/dataflow/src/sink/tail.rs:45]: (If the answer is: it's a side-effect of #5028 then that's good enough. I don't know how sinks work and that's fine.)

Comment (umanwizard) [reply src/dataflow/src/sink/tail.rs:45]: Yes, it is guaranteed upstream, when the differential-dataflow diff is moved into the data.

**Comment by frankmcsherry** [comment]
Personal take is that these changes leave us with a function that is easy to mis-use. Idk what the constraints are elsewhere in the code, but this seems like if anything rots around it (e.g. field order shuffled, multiplicities presented) we get to find out about it (maybe) when it goes wrong, rather than ahead of time.

**Comment by benesch** [comment]
> Personal take is that these changes leave us with a function that is easy to mis-use. Idk what the constraints are elsewhere in the code, but this seems like if anything rots around it (e.g. field order shuffled, multiplicities presented) we get to find out about it (maybe) when it goes wrong, rather than ahead of time.

The function is already exactly this hard to misuse. The order of fields in the `Row` is already assumed if `connector.emit_progress` is true. And the output of `TAIL` will not match its documented behavior if multiplicities are presented.

I also find this unfortunate, but upgrading this assertion makes this precondition _more_ clear, not less, in my opinion.

**Comment by benesch** [comment]
Converting to a draft because this review stalls out, but I'll note that I don't think anyone has satisfactorily rebutted my points in my last comment.

**Comment by benesch** [comment]
This code is long gone.

**Pull request closed without merging**